internal/devpkg/narinfo_cache.go

```diff
@@ -103,39 +65,44 @@ func FillNarInfoCache(ctx context.Context, packages ...*Package) error {
 
 	group, _ := errgroup.WithContext(ctx)
 	for _, p := range eligiblePackages {
-		// If the package's NarInfo status is already known, skip it
-		isNarInfoInCache.lock.RLock()
-		_, ok := isNarInfoInCache.status[p.Raw]
-		isNarInfoInCache.lock.RUnlock()
-		if ok {
-			continue
-		}
 		pkg := p // copy the loop variable since its used in a closure below
 		group.Go(func() error {
-			err := pkg.fillNarInfoCache()
-			if err != nil {
-				// default to false if there was an error, so we don't re-try
-				isNarInfoInCache.lock.Lock()
-				isNarInfoInCache.status[pkg.Raw] = false
-				isNarInfoInCache.lock.Unlock()
-			}
+			_, err := pkg.fetchNarInfoStatusOnce()
 			return err
 		})
 	}
 	return group.Wait()
 }
 
-// fillNarInfoCache fills the cache value for the narinfo of this package,
-// assuming it is eligible for the binary cache. Callers are responsible
-// for checking isEligibleForBinaryCache before calling this function.
-//
-// NOTE: this must be concurrency safe.
-func (p *Package) fillNarInfoCache() error {
+// narInfoStatusFnCache contains cached OnceValues functions that return cache
+// status for a package. In the future we can remove this cache by caching
+// package objects and ensuring packages are shared globally.
+var narInfoStatusFnCache = sync.Map{}
+
+// fetchNarInfoStatusOnce is like fetchNarInfoStatus, but will only ever run
+// once and cache the result.
+func (p *Package) fetchNarInfoStatusOnce() (bool, error) {
+	type inCacheFunc func() (bool, error)
+	f, ok := narInfoStatusFnCache.Load(p.Raw)
+	if !ok {
+		f = inCacheFunc(sync.OnceValues(func() (bool, error) {
+			return p.fetchNarInfoStatus()
+		}))
+		f, _ = narInfoStatusFnCache.LoadOrStore(p.Raw, f)
+	}
```

Comment:
We could replace this with a single LoadOrStore but that would create throwaway OnceValues function every time.